### ISSUE TITLE
v 6.1.0 with Tor + Electrum hangs

### ISSUE THREAD

**Issue by Darth-Coin** [BlueWallet/BlueWallet#3079]
Android 10, BW 6.1.0 recently updated
Had some wallets already (onchain and LN).
Wanted to test a connection to my Electrum server, from my Umbrel node.
This morning, setup BW with my electrum server behind Tor, all was connected OK, synced in few seconds, using Wifi.
After several hours, not using BW, went out etc, at home back opening the BW, shows wallets in "updating".
I thought OK, tor is making funny things, wanted ot reset connection to default. I could enter in settings, but hangs in that moment.
No move, no click no nothing.
Force closed the app. Restarted. Hanged again in main screen updating.
I can slide the wallets, I can browse history (can't click on them), but nothing else.
Tried to switch from wifi to 4G, force closed again, waited a bit but still hanged.

Wen to check my Electrum server, testing with Electrum desktop, connected through Tor pointing to my node: worked perfect
Tested Electrum with BW 6.0.08 in another mobile (using Orbot) and pointing to my electrum server: worked.

If you need logs from my node, let me know.

It should have a timeout or something if the Tor connection cannot find a good connection and give the user an error saying that connection lost or something and be able de reset to default. or even the message to have an option like "do you want to reset to default?" or " do you want to refresh the connection?"

**Comment by karozagorus** [comment]
Same stuff is happening to me! Android 11 here.

**Comment by Overtorment** [comment]
so, basically, tor on android hangs sometimes? but not on ios?

**Comment by Darth-Coin** [comment]
@Overtorment not sometimes, but only if you update to v 6.1.1 and use tor + electrum server
If you use v 6.1.1 normally with BW servers, it's just fine.
If you use v6.1.1 with Orbot, is just fine
If you use v 6.0.8 + Electrum + Orbot , is just fine
If you use v 6.0.8 alone, BW servers is just fine
If you use v6.0.8 + Tor + Electrum + Umbrel Lndhub, is just fine
If you use v6.1.1 + Tor in app + Electrum + Umbrel Lndhub, it hangs, can't even enter in network settings to revert the changes. You have to uninstall and install again

**Comment by Overtorment** [comment]
Ok, tor in app on android hangs

**Comment by Overtorment** [comment]
Can you pinpoint what causes freeze, tor lndhub or tor electrum?

**Comment by gabidi** [comment]
@Darth-Coin  @karozagorus When BW freezes does restarting BlueWallet resolve the issue for a bit ? I mean can you connect/use your wallets while connected to your nodes over Tor before it hangs again ? Or the only way to have it work again is to uninstall BW and reinstall it ?

**Comment by Darth-Coin** [comment]
@gabidi I did this:
- force close BW app
- open again and immediately freezes, once I tried to be fast and enter into settings, but was freezing again.
- I tried also to leave it for a while doing his job, but no it stuck
The only way to make it work was uninstall the app, install fresh, restore wallets with default settings

**Comment by gabidi** [comment]
@Darth-Coin Ok that's helpful thanks.

**Comment by gabidi** [comment]
![bw_test](https://user-images.githubusercontent.com/315931/117448802-2da29f00-af0d-11eb-9c99-7e5172c654dd.gif)
I've had BW running in a simulator connected to an Electrum node and LND hub via Tor for almost 24 hours now and I can't seem to seem re replicate the error above...

@Darth-Coin  @karozagorus  Would you mind letting me know what the devices you where running when BW on when the issue above happened ? 
I'll try this weekend by running on a physical Android 11 device (Only one I have is running Graphene but i think that should be ok) .

**Comment by Darth-Coin** [comment]
@gabidi I have these 2 phones (not virtual ones):
- LG Android 8.1, stock ROM (clean as possible from garbage)
- Motorola Android 10 stock ROM (my wife's mobile)
As I said, the old one Android 8.1 + BW 6.0.8 + Orbot works fine. If I update to 6.1.1 and use integrated Tor it freeze. What I do not test is to update to 6.1.1  not use Tor integrated but with orbot.
The other one android 10 + integrated tor, freeze. Tried to restart the app and is doing the same. In the end I reverted to 6.0.8 so my wife can still use BW.

I can't say how is on GrapheneOS because these mobiles do not support it (I wish I could).
I have another mobile with LineageOS 14 but is too old to work with BW 6.1.1.

**Comment by briheim** [comment]
I have exact same issue after setting up my own electrum server in BW android app. Just personal lightning node with lnhub was running without problems.

Android 10
OxygenOS 10.3.10
BW Android 6.1.1
Umbrel node 0.3.10 running electrum and lnhub

**Comment by ncoelho** [comment]
Manage to replicate the freeze with a S20 Ultra, android 10 and 11.

You go to any view, and after that it just freezes.


<img width="1437" alt="Screenshot 2021-05-20 at 15 33 25" src="https://user-images.githubusercontent.com/1238429/118987819-e7bfef00-b980-11eb-8e0d-a29006dd2bd6.png">

<img width="1437" alt="Screenshot 2021-05-20 at 15 39 23" src="https://user-images.githubusercontent.com/1238429/118988535-906e4e80-b981-11eb-96f4-ad8333f4992e.png">

**Comment by ncoelho** [comment]
@gabidi can give you browserstack access, real devices on demand with some debug info. Just send me your email n@bluewallet.io

**Comment by Overtorment** [comment]
> @gabidi can give you browserstack access, real devices on demand with some debug info. Just send me your email [n@bluewallet.io](mailto:n@bluewallet.io)

can it be replicated with local emulator  (S20/android11) ?

**Comment by ncoelho** [comment]
> > @gabidi can give you browserstack access, real devices on demand with some debug info. Just send me your email [n@bluewallet.io](mailto:n@bluewallet.io)
> 
> can it be replicated with local emulator (S20/android11) ?

Not really, at least doesn't freeze immediately and has not freeze yet.

**Comment by gabidi** [comment]
@ncoelho  @Overtorment  I managed to get it to freeze on Pixel 5 running Graphene last week and started to look into it.
Will be continuing this weekend and update you..

**Comment by Overtorment** [comment]
Awesome!

**Comment by ncoelho** [comment]
Another behaviour on a pixel 4 android 11. The app just loads forever on the splashing screen without passing this view with Tor enabled.

**Comment by gabidi** [comment]
@ncoelho  That might be a function of how and when Bluewallet initiates a Tor connection and awaits it's completion.
If Tor fails and takes too long to connect/establish a circuit and Bluewallet waits for Tor to connect before proceeding you would see that. 
There's a connection timeout (https://github.com/Sifir-io/react-native-tor/blob/main/src/index.tsx#L354) setting that can be passed where the module will throw an error if initializing Tor takes longer than supplied. At which point you can inform the user to retry connecting etc..

**Comment by Overtorment** [comment]
> @ncoelho That might be a function of how and when Bluewallet initiates a Tor connection and awaits it's completion.
> If Tor fails and takes too long to connect/establish a circuit and Bluewallet waits for Tor to connect before proceeding you would see that.
> There's a connection timeout (https://github.com/Sifir-io/react-native-tor/blob/main/src/index.tsx#L354) setting that can be passed where the module will throw an error if initializing Tor takes longer than supplied. At which point you can inform the user to retry connecting etc..

but the default timeout is 25 sec, right? so just waiting for 26 seconds user should have seen some response. 
but I gotta check if we have this call wrapped in try-catch

**Comment by gabidi** [comment]
@Overtorment Yep default is 25 seconds, So should error out after.

**Comment by vindard** [comment]
I just had this happen to me as well on Android. I had 2 of the same Lightning wallets in my app, 1 tied to my local clearnet address on my LAN for my Pi and the other pointed at the Tor address.

If Orbot is off and I open the app, I noticed the app freezes after a few seconds no matter what sequence of things I do. If I remove the Tor wallet and leave the clearnet one alonenthen the issue goes away.

That's the extent of my troubleshooting, but I'm wondering if it could be something with a bad sync via Tor if Tor isn't connected/reachable for some reason?

**Comment by ncoelho** [comment]
@gabidi any progress on this issue? A lot of users using tor, which is great, but dealing with a lot of complains as well, like daily :/

Let us know if help is needed.

**Comment by gabidi** [comment]
@ncoelho Still stumped by issue. 
Debugged quite a bit and found some small bugs here and there, which where fixed and improved efficiency etc and Bluewallet still hangs when using the default JS engine. 
For some reason RN's  rendering thread just silently fails with no errors in logs etc. Note that the main thread still works fine and that's why Scroll views still work even after this issues happens (Scroll views run on the main thread). 
At this time the solution that seems to work is to change to RN V8 which runs smoothly.

Site note: @Overtorment Have you tried your `rn-ldk` lib in BW ? Curious if you get freezes since the stack is very similar.

Anyway, I'll keep looking into this and let you know when I find a solution that doesn't involve V8.

**Comment by Overtorment** [comment]
@gabidi its weird that a lot of other rn modules are not causing any issues, most notoriously `react-native-tcp-socket` which provides tcl & tls sockets, similar to tor package.
I'm currently integrating `rn-ldk` and don't see any freezes, for now. same when I was working on integrating `rn-lnd` which is also based on the same RN module template  (react-native-builder-bob)

i think the only thing left for you to debug is to reliably reproduce the issue, and then start removing pieces of code till the issue is gone. that's how you'll be able to pinpoint location of the rootcause. 
or vice versa, start from scratch and add code till  the issue appears...


PS. i wonder if switching to hermes engine will fix the issue.

**Comment by ncoelho** [comment]
Can anyone with this issue try the latest .apk on this link and report back please

https://github.com/BlueWallet/BlueWallet/pull/3170#issuecomment-874354957

**Comment by Darth-Coin** [comment]
Just tested on Android 10 (Motorola G Plus), stack OS.
- removed previous BW
- install this testing vers.
- didn't create/restore any wallet
- go to settings - network - tor, activate and test. OK
- go to electrum server, scan my Umbrel node QR, connected
- clicked the back arrow, app is frozen
- force closed, re-open, check the tor, OK, but when I clicked back it freeze

I can test also on Android 8.1 but I think at this moment will not be useful and not sure if is supported on android 8

**Comment by ncoelho** [comment]
> Just tested on Android 10 (Motorola G Plus), stack OS.
> 
> * removed previous BW
> * install this testing vers.
> * didn't create/restore any wallet
> * go to settings - network - tor, activate and test. OK
> * go to electrum server, scan my Umbrel node QR, connected
> * clicked the back arrow, app is frozen
> * force closed, re-open, check the tor, OK, but when I clicked back it freeze
> 
> I can test also on Android 8.1 but I think at this moment will not be useful and not sure if is supported on android 8

Don't touch on the Tor tab, that is a debug windown, not to activate manually

**Comment by bezysoftware** [comment]
Not sure if it's a coincidence, but I was experiencing freezes on almost every start up, but once I disabled "View all wallets" in on launch settings, it hardly ever happens anymore. 
Pixel 5, Android 11

**Comment by damanic** [comment]
Same issue on Android 11 Pixel 4A.

Effectively unusable, freezes up within a few touch interactions.

Is this likely to be fixed in next release?

**Comment by rkfg** [comment]
Same here on Moto G7 Power, Android 10. Added two LN wallets using my own LndHub installation (and my own node of course). The one that uses a public URL works fine, another wallet that connects to the same instance but via Tor (I setup a hidden .onion service for it) works for a short while, then the whole app just freezes randomly. I can only do 1-2 transactions, then I have to swipe the app away and restart. After I deleted the .onion wallet the app stopped freezing.

It's very sad because so many people don't have WAN IPs at home and not many have their own VPSes to forward ports (like I did). Tor provides global availability to the devices behind NAT, it's sometimes even more important than privacy and encryption. And as I understand it's not possible to disable the built-in Tor and use Orbot instead because the built-in version starts automatically as needed.

**Comment by Overtorment** [comment]
@rkfg you should be able to use orbot, theoretically. just don't provide any `.onion` URLs to bluewallet (electrum or lndhub) and it wont start tor daemon

**Comment by rkfg** [comment]
@Overtorment yes, but that's the point. If I can't expose my node and services directly to connect to them via the internet I need to use some kind of proxy/VPN and Tor is the most frictionless solution. Anything else requires a server with a dedicated IP.

**Comment by psoetens** [comment]
I had also bricked the app by setting the lndhub and electrum server of Umbrel. After resetting the app a dozen times to no avail (immediate lock-ups), I decided to clear the BlueWallet App cache in Android, and it immediately worked again. It allowed me to change the settings to defaults again. 

So could others confirm that clearing the app cache also resolves the issue for them?

**Comment by rkfg** [comment]
Clearing the cache didn't solve the issue for me. I was renaming my wallets and it froze just after saving the changes. 
![Screenshot_20210821-135413](https://user-images.githubusercontent.com/184066/130319578-1827531c-2857-4cdc-bdc9-7815a3c77e99.png)
This spinner was spinning indefinitely, so this freeze isn't full. At least the UI is being redrawn. It rather looks like the app loses connection with some other underlying system (or some thread deadlocks) so the UI stops reacting to events.

Prior to this I was also renaming my wallets and I noticed that the save button didn't react to taps. However, I was still able to type in the wallet name field. When I confirmed it using the keyboard (the green checkmark key that replaces the Enter key during editing) this text field lost focus and nothing else could have been tapped, no control reacted. So I had to swipe the app off and restart it.

**Comment by rkfg** [comment]
Easy way to trigger this:
1) Run the app anew, don't wait until it synchronizes the wallets
2) Open an LN wallet
3) Open its properties (three dots in the top right corner)
4) Toggle "Display in Wallets List"
5) Tap Save
6) Repeat this process quickly

With Tor turned off I did this 10 times with no problems. With Tor active I could do this only 4 times and it hanged during saving with the spinner spinning forever. Another time the Save button failed to even appear lol. The UI is partially responsive, this toggle shows ripple animation when I tap it but it doesn't switch actually. Nothing else reacts to taps.

If the app is left alone for a while with Tor running it seems that it stops freezing this easily or it requires more actions. If you start changing wallets properties right after the main activity appears it should be easy to lock it up.

**Comment by Overtorment** [comment]
while we cant fix tor freezes on android directly, we are anticipating several dependencies to be updated soon which might resolve tor freezes for good, so bear with us.

**Comment by damanic** [comment]
Updated to 6.2.3  - Unfortunately this issue persists 

- Removed APP and re-installed.
- Went to settings -> Tor -> start
- Application froze/crashed.

Fixing this should be a priority IMO. The application continues to be unusable on android for anyone that needs to interact with TOR nodes.

I have tried with Orbot as VPN for BlueWallet, but it appears the BlueWallet inbuilt TOR service is in conflict. 

If the BlueWallet TOR deamon had a toggle that allowed it to be disabled entirely, would this allow us to use Orbot instead?

**Comment by Overtorment** [comment]
with orbot, can you just specify `localhost` as a connection host?
in that case internal Tor wont kick in since it doesn't have `.onion` in the string

**Comment by damanic** [comment]
@Overtorment 
I'm not sure how a localhost address would work in this case.  Are you suggesting I use a clearnet proxy to connect BlueWallet to a electrum/lightning node that is only accessible over TOR?  

If the TOR deamon built into BlueWallet could be toggled off, do you think Orbot VPN would work in its place?  This would be a work around most people could set up without too much difficulty.

**Comment by Overtorment** [comment]
not exactly sure how orbot works. i thought it provides some kind of port forwarding on localhost so any app can connect and use tor.
we had people use robot cusscessfully before we added embedded tor daemon.

and yes, tor can be disabled. just don't put any hosts (or lndhub URLs) in settings that contain `.onion`, and restart

**Comment by damanic** [comment]
Posting my workaround in case it helps those following.

I have got BlueWallet Android connected to my TOR ONLY Umbrel Node via a self hosted HTTPS -> TOR proxy, and it is STABLE!

I have published a docker for the proxy here: https://hub.docker.com/r/damanic/socatorssl

The example exposes only the Electrum Server port to clearnet, but you can run multiple instances of the proxy to expose other service ports too  (Eg. LNDHub)

**Comment by fightlessbirds** [comment]
If I use clearnet for Electrum and onion for Lightning, it works. The freeze only happens for me if I use onion for both Electrum and Lightning.

**Comment by pjw65** [comment]
BlueWallet 6.2.17 reezes/crashes on Android 11 when using custom electrum server (umbrel) and/or LND Hub over internal tor.

Workaround: 
- Settings/Network/Tor disable
- Orbot VPN (add Apps/BlueWallet)

**Comment by saerdnaa** [comment]
Will there ever be a solution?
